enforcer-rules/src/main/java/com/google/cloud/tools/opensource/enforcer/LinkageCheckerRule.java

```diff
@@ -0,0 +1,158 @@
+import com.google.cloud.tools.opensource.classpath.ClasspathCheckReport;
+import com.google.cloud.tools.opensource.classpath.ClasspathChecker;
+import com.google.cloud.tools.opensource.classpath.JarLinkageReport;
+import com.google.cloud.tools.opensource.dependencies.RepositoryUtility;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+import javax.annotation.Nonnull;
+import org.apache.maven.RepositoryUtils;
+import org.apache.maven.enforcer.rule.api.EnforcerRule;
+import org.apache.maven.enforcer.rule.api.EnforcerRuleException;
+import org.apache.maven.enforcer.rule.api.EnforcerRuleHelper;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.plugin.logging.Log;
+import org.apache.maven.project.DefaultDependencyResolutionRequest;
+import org.apache.maven.project.DependencyResolutionException;
+import org.apache.maven.project.DependencyResolutionRequest;
+import org.apache.maven.project.DependencyResolutionResult;
+import org.apache.maven.project.MavenProject;
+import org.apache.maven.project.ProjectDependenciesResolver;
+import org.codehaus.plexus.component.configurator.expression.ExpressionEvaluationException;
+import org.codehaus.plexus.component.repository.exception.ComponentLookupException;
+import org.eclipse.aether.RepositoryException;
+import org.eclipse.aether.RepositorySystemSession;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.graph.Dependency;
+
+/** Linkage Checker Maven Enforcer Rule. */
+public class LinkageCheckerRule implements EnforcerRule {
+
+  /** Set to rue if the rule is used for a BOM project. */
+  private boolean bom = false;
+
+  @Override
+  public void execute(@Nonnull EnforcerRuleHelper helper) throws EnforcerRuleException {
+    Log log = helper.getLog();
+
+    try {
+      MavenProject project = (MavenProject) helper.evaluate("${project}");
+      MavenSession session = (MavenSession) helper.evaluate("${session}");
+      RepositorySystemSession repositorySystemSession = session.getRepositorySession();
+
+      ImmutableList<Path> classpath =
+          isBomProject(project)
+              ? findBomClasspath(project)
+              : findProjectClasspath(project, repositorySystemSession, helper);
+
+      List<Path> artifactJarsInProject = classpath.subList(0, project.getDependencies().size());
+      ImmutableSet<Path> entryPoints = ImmutableSet.copyOf(artifactJarsInProject);
+
+      try {
+        ClasspathChecker classpathChecker = ClasspathChecker.create(classpath, entryPoints);
+        ClasspathCheckReport linkageReport = classpathChecker.findLinkageErrors();
+        int totalErrors =
+            linkageReport.getJarLinkageReports().stream()
+                .mapToInt(JarLinkageReport::getCauseToSourceClassesSize)
+                .sum();
+        if (totalErrors > 0) {
+          log.info(
```

Comment:
This should be log.error if it's failing the build.

I do think the rule needs a configuration option to specify the severity of these errors.